zcash_client_sqlite/src/wallet.rs

```diff
@@ -727,13 +741,37 @@ pub(crate) fn get_wallet_summary<P: consensus::Parameters>(
                 balances.add_unshielded_value(value)?;
             }
         }
+        drop(transparent_trace);
     }
 
+    let next_sapling_subtree_index = {
+        // This will return a runtime error if we call `get_wallet_summary` from two
+        // threads at the same time, as transactions cannot nest.
+        let tx = conn.unchecked_transaction()?;
+        let shard_store =
+            SqliteShardStore::<_, ::sapling::Node, SAPLING_SHARD_HEIGHT>::from_connection(
+                &tx,
+                SAPLING_TABLES_PREFIX,
+            )?;
+
+        // The last shard will be incomplete, and we want the next range to overlap with
+        // the last complete shard, so return the index of the second-to-last shard root.
+        shard_store
+            .get_shard_roots()
```

Comment:
Instead of constructing a `SqliteShardStore` and calling `get_shard_roots`, I think it would be better to just query the `sapling_shard_roots` table directly. It _should_ be equivalent, but querying `sapling_shard_roots` directly ensures that we _only_ get data downloaded from lightwallet, and not internally derived roots.

Reply:
I'm confused; I cannot find any `sapling_shard_roots` table. There is only `sapling_tree_shards`, which is where `WalletCommitmentTrees::put_sapling_subtree_roots` inserts into, and what `shard_store.get_shard_roots()` fetches from.

Reply:
Oops, yes, `sapling_tree_shards` is what I meant. 

One minor improvement would be to use https://github.com/zcash/librustzcash/blob/main/zcash_client_sqlite/src/wallet/commitment_tree.rs#L481 directly, rather than having to instantiate a full `SqliteShardStore`.

Reply:
Eh, `SqliteShardStore::from_connection` is literally just returning a container without any extra data storage or logic going on, so it's likely that the compiler will optimise this out. I'm fine with the more explicit type-ing.

Reply:
That's okay, I guess. It's a little less insulated against future changes to SaplingShardStore construction; I'd still slightly prefer it the way I suggested but I don't mind too much.